[appearance: base] Add -internal-auto-base() and resolve it by substitution

shallawa opened the pull request:
#### 04ea7c36cab8de930b1477f15e9fc921380e080c
<pre>
[appearance: base] Add -internal-auto-base() and resolve it by substitution
<a href="https://bugs.webkit.org/show_bug.cgi?id=303078">https://bugs.webkit.org/show_bug.cgi?id=303078</a>
<a href="https://rdar.apple.com/165381349">rdar://165381349</a>

Reviewed by Tim Nguyen.

To add support for `appearance: base` a new internal function `-internal-auto-base()`
is added. This functional will resolve and substitute itself with one of two values
depending on whether the style has `appearance: base` or not.

No UA style sheet changes will be added in this PR. An internal setting which
enables parsing `-internal-auto-base()` on regular author style will be added for
now. This setting will be used to write layout tests until default properties for
`appearance: base` form controls are added in the UA style sheet.

Test: fast/forms/appearance-base/appearance-base-checkable-basics.html

* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-basics-expected.html: Added.
* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-basics.html: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/css/CSSProperty.h:
(WebCore::CSSProperty::CSSProperty):
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::applyUASheetBehaviorsToContext):
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeFunctionArgument):
(WebCore::consumeInternalAutoBaseFunction):
(WebCore::consumeStyleProperty):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Primitives.cpp:
(WebCore::CSSPropertyParserHelpers::consumeArgument):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Primitives.h:
* Source/WebCore/css/parser/CSSPropertyParserResult.cpp:
(WebCore::CSS::PropertyParserResult::addProperty):
* Source/WebCore/css/parser/CSSPropertyParserResult.h:
* Source/WebCore/css/parser/CSSVariableParser.cpp:
(WebCore::isValidDashedFunction):
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyProperty):
(WebCore::Style::Builder::resolveInternalAutoBaseFunction):
* Source/WebCore/style/StyleBuilder.h:

Canonical link: <a href="https://commits.webkit.org/303693@main">https://commits.webkit.org/303693@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3be4e48335adf0568ffb61842f16575cd579adaf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133371 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5872 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44507 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140926 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85418 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1e2e5315-1517-4bc0-b044-e611c4afc0b6) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135241 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6385 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5737 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/102029 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85418 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a563f111-3f9c-41ee-898f-27fc174d0469) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136318 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4530 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119555 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82820 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c7df78ba-1da2-4c2a-8a54-29f3574d24fb) 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/132720 "Build is in progress. Recent messages:") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/4409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/2004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/125443 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113504 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37663 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143573 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/131883 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5542 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38249 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/110403 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/132799 "Build is in progress. Recent messages:") | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/5624 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/4763 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110582 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28014 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115810 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59292 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5597 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/34145 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/164847 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5443 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69049 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43070 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5686 "Built successfully") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/5553 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -226,6 +227,15 @@ static std::optional<CSSWideKeyword> consumeCSSWideKeyword(CSSParserTokenRange&
+
+static bool consumeCSSFunctionValue(CSSParserTokenRange& range, CSSPropertyID property, CSS::PropertyParserState& state, CSS::PropertyParserResult& result)
+{
+    if (range.peek().functionId() == CSSValueInternalAutoBase)
```
I think you'll need to gate this to `state.context.mode == UASheetMode` to avoid web-exposing this.

shallawa replied:
This condition was added.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Appearance.cpp:
```diff
@@ -0,0 +1,98 @@
+    return true;
+}
+
+static bool consumeFunctionPatameter(CSSParserTokenRange& range, CSSPropertyID property, CSS::PropertyParserState& state, CSS::PropertyParserResult& result)
```
```suggestion
static bool consumeFunctionParameter(CSSParserTokenRange& range, CSSPropertyID property, CSS::PropertyParserState& state, CSS::PropertyParserResult& result)
```

shallawa replied:
Spelling was fixed.

anttijk commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -226,6 +227,15 @@ static std::optional<CSSWideKeyword> consumeCSSWideKeyword(CSSParserTokenRange&
 
+// MARK: - function value consumer
+
+static bool consumeCSSFunctionValue(CSSParserTokenRange& range, CSSPropertyID property, CSS::PropertyParserState& state, CSS::PropertyParserResult& result)
```
This should be called something more specific like `consumeInternalAutoBaseFunction`.

shallawa replied:
Function name was changed.

anttijk commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -621,6 +631,9 @@ bool consumeStyleProperty(CSSParserTokenRange& range, const CSSParserContext& co
         }
 
+        if (consumeCSSFunctionValue(range, property, state, result))
+            return true;
```
I think this call can be earlier, before `if (WebCore::isShorthand(property))` test so you don't need two call sites.

shallawa replied:
This statement was moved and it is now called once.

anttijk commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Appearance.cpp:
```diff
@@ -0,0 +1,98 @@
+        const auto& token = range.consumeIncludingWhitespace();
+        if (token.type() == FunctionToken && !consumeFunctionBody(range))
+            break;
+    }
```
I'm not sure what is the best way to write the parsing code for this with all the new CSS parsing infra but @weinig surely has ideas. 

I would do something more like the code in `isValidDashedFunction()` which is similar in that it parses a comma separated argument list of arbitrary CSS values. The `validateArgument` lambda would be replaced with the code to actually parse the argument ranges.

shallawa replied:
The lambda `consumeArgument()` was moved to a new function called `CSSPropertyParserHelpers::consumeArgument()`. A new argument called `index` is added to force consuming a comma if its value is greater than zero. This function is now called from `isValidDashedFunction()` and `consumeFunctionArgument()`.

anttijk commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Appearance.h:
```diff
@@ -0,0 +1,44 @@
+
+namespace CSSPropertyParserHelpers {
+
+bool consumeAutoBaseFunction(CSSParserTokenRange&, CSSPropertyID, CSS::PropertyParserState&, CSS::PropertyParserResult&);
```
A file called PropertyParserConsumer+Appearance.h seems like wrong place to put this code. The code here is about parsing -internal-auto-base(), not the appearance property.

shallawa replied:
This file was removed. `consumeInternalAutoBaseFunction()` was added in CSSPropertyParser.cpp.

anttijk commented on Source/WebCore/html/BaseCheckableInputType.cpp:
```diff
@@ -134,4 +139,28 @@ void BaseCheckableInputType::fireInputAndChangeEvents()
+}
+
+Ref<HTMLDivElement> BaseCheckableInputType::createShadowSubtreeElement(const AtomString& userAgentPart)
+{
```
It would be better to split out all DOM related things and just focus on the style mechanism `-internal-auto-base()` with this patch.

shallawa replied:
DOM changes were removed.

anttijk commented on Source/WebCore/css/html.css:
```diff
@@ -1422,3 +1422,46 @@ ins, del {
+        width: -internal-auto-base(unset, 2em);
+        border: -internal-auto-base(unset, 0.3em solid currentColor);
+        background-color: -internal-auto-base(unset, transparent);
+    }
```
Lets not do UA style sheet changes in this patch.

While this value should not be allowed outside the UA sheet (as Tim pointed out) you could add an internal setting that enabled it on regular author style (which becomes a bit in CSSParserContext) that enables it anyway. This way tests can be written easily as regular layout tests.

shallawa replied:
UA style sheet changes were removed. A new internal setting called `CSSInternalAutoBaseParsingEnabled` to allow parsing `-internal-auto-base()` in the author style sheet.

anttijk commented on Source/WebCore/style/StyleBuilder.cpp:
```diff
@@ -546,6 +547,22 @@ void Builder::applyCustomProperty(const AtomString& name, Variant<Ref<const Styl
     );
 }
 
+Ref<CSSValue> Builder::resolveFunctionReferences(CSSPropertyID propertyID, CSSValue& value)
```
This should have more specific name like `resolveInternalAutoBaseFunction`

shallawa replied:
Done.

anttijk commented on Source/WebCore/style/StyleBuilder.cpp:
```diff
@@ -546,6 +547,22 @@ void Builder::applyCustomProperty(const AtomString& name, Variant<Ref<const Styl
+    if (!result)
+        return value;
+
+    return resolveFunctionReferences(propertyID, *result);
```
I don't think there is need for this recursion. There is no reason we would ever nest `-internal-auto-base()`. We could assert against it though. 

Parsing code also does not need to handle nesting.

shallawa replied:
Recursion was removed.

nt1m commented on LayoutTests/fast/forms/appearance-base/appearance-base-checkable-basics.html:
```diff
@@ -0,0 +1,70 @@
+</style>
+<script>
+    if (window.internals)
+        internals.setCSSTextTransformMathAutoEnabled(true);
```
You can just do `<!-- webkit-test-runner [ CSSInternalAutoBaseParsingEnabled=true ] -->` without any internals hooks.

shallawa replied:
Done.

nt1m commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -1238,6 +1238,20 @@ CSSInputSecurityEnabled:
+  type: bool
+  status: internal
+  category: css
+  humanReadableName: "CSS appearance: -internal-auto-base()"
```
```suggestion
  humanReadableName: "CSS -internal-auto-base() in author sheet"
```

shallawa replied:
Done.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -226,6 +227,64 @@ static std::optional<CSSWideKeyword> consumeCSSWideKeyword(CSSParserTokenRange&
+
+static bool parseInternalAutoBaseFunction(CSSParserTokenRange& range, CSSPropertyID property, CSS::PropertyParserState& state, CSS::PropertyParserResult& result)
+{
+    if (!(state.context.mode == UASheetMode || state.context.cssInternalAutoBaseParsingEnabled))
```
```suggestion
    if (state.context.mode != UASheetMode && !state.context.cssInternalAutoBaseParsingEnabled)
```

nt1m replied:
For the first check, you could actually also just extend `applyUASheetBehaviorsToContext` instead of adding the first check.

nt1m commented on LayoutTests/fast/forms/appearance-base/appearance-base-checkable-basics.html:
```diff
@@ -0,0 +1,47 @@
+    </div>
+    <div class="disabled">
+        <input type="checkbox" disabled />
+        <input type="radio" disabled />
```
```suggestion
        <input type="checkbox">
        <input type="radio">
    </div>
    <div class="disabled">
        <input type="checkbox" disabled>
        <input type="radio" disabled>
```

nt1m commented on Source/WebCore/css/parser/CSSParserContext.cpp:
```diff
@@ -160,7 +162,10 @@ void add(Hasher& hasher, const CSSParserContext& context)
+
+    uint32_t bits2 = context.cssInternalAutoBaseParsingEnabled;
+
+    add(hasher, context.baseURL, context.charset, context.propertySettings, context.mode, bits, bits2);
```
We could probably clean up some bits to avoid adding that. I can take care of that!

nt1m reviewed (approved):
Looks good to me, would be good if @anttijk could take a look before landing!

annevk commented on Source/WebCore/css/CSSValueKeywords.in:
```diff
@@ -929,6 +929,7 @@ attachment enable-if=ENABLE_ATTACHMENT_ELEMENT
 borderless-attachment enable-if=ENABLE_ATTACHMENT_ELEMENT
 textarea
 textfield
+-internal-auto-base
```
I still have a mild preference for calling this `-internal-appearance`. While that doesn't convey what the arguments correspond to, I think it is a more clear in that at least you know what feature this relates to. (With `light-dark` that problem doesn't exist as the feature is called "dark mode".)

nt1m commented on Source/WebCore/css/parser/CSSParserContext.cpp:
```diff
@@ -157,6 +159,7 @@ void add(Hasher& hasher, const CSSParserContext& context)
         | context.cssConstrainedDynamicRangeLimitEnabled    << 28
         | context.cssTextDecorationLineErrorValues          << 29
         | context.cssTextTransformMathAutoEnabled           << 30;
+        | context.cssInternalAutoBaseParsingEnabled         << 31;
```
```suggestion
        | context.cssTextTransformMathAutoEnabled           << 30
        | context.cssInternalAutoBaseParsingEnabled         << 31;
```

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -226,6 +227,64 @@ static std::optional<CSSWideKeyword> consumeCSSWideKeyword(CSSParserTokenRange&
+        return false;
+    return consumeInternalAutoBaseFunction(range, property, state, result);
+}
+
```
Do we need both `parseInternalAutoBaseFunction` and `consumeInternalAutoBaseFunction` ? Can they just be a single method?

shallawa committed:
[appearance: base] Add -internal-auto-base() and resolve it by substitution
https://bugs.webkit.org/show_bug.cgi?id=303078
rdar://165381349

Reviewed by Tim Nguyen.

To add support for `appearance: base` a new internal function `-internal-auto-base()`
is added. This functional will resolve and substitute itself with one of two values
depending on whether the style has `appearance: base` or not.

No UA style sheet changes will be added in this PR. An internal setting which
enables parsing `-internal-auto-base()` on regular author style will be added for
now. This setting will be used to write layout tests until default properties for
`appearance: base` form controls are added in the UA style sheet.

Test: fast/forms/appearance-base/appearance-base-checkable-basics.html

* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-basics-expected.html: Added.
* LayoutTests/fast/forms/appearance-base/appearance-base-checkable-basics.html: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/css/CSSProperty.h:
(WebCore::CSSProperty::CSSProperty):
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::applyUASheetBehaviorsToContext):
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeFunctionArgument):
(WebCore::consumeInternalAutoBaseFunction):
(WebCore::consumeStyleProperty):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Primitives.cpp:
(WebCore::CSSPropertyParserHelpers::consumeArgument):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Primitives.h:
* Source/WebCore/css/parser/CSSPropertyParserResult.cpp:
(WebCore::CSS::PropertyParserResult::addProperty):
* Source/WebCore/css/parser/CSSPropertyParserResult.h:
* Source/WebCore/css/parser/CSSVariableParser.cpp:
(WebCore::isValidDashedFunction):
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyProperty):
(WebCore::Style::Builder::resolveInternalAutoBaseFunction):
* Source/WebCore/style/StyleBuilder.h:

Canonical link: https://commits.webkit.org/303693@main
